src/interp/binary-reader-interp.cc

```diff
@@ -1394,6 +1500,110 @@ Result BinaryReaderInterp::OnTableInitExpr(Index segment_index,
   return Result::Ok;
 }
 
+Result BinaryReaderInterp::OnThrowExpr(Index tag_index) {
+  CHECK_RESULT(validator_.OnThrow(loc, Var(tag_index)));
+  istream_.Emit(Opcode::Throw, tag_index);
+  return Result::Ok;
+}
+
+Result BinaryReaderInterp::OnRethrowExpr(Index depth) {
+  Index catch_depth;
+  CHECK_RESULT(validator_.OnRethrow(loc, Var(depth)));
+  CHECK_RESULT(validator_.GetCatchCount(depth, &catch_depth));
+  // The rethrow opcode takes an index into the exception stack rather than
+  // the number of catch nestings, so we subtract one here.
+  istream_.Emit(Opcode::Rethrow, catch_depth - 1);
+  return Result::Ok;
+}
+
+Result BinaryReaderInterp::OnTryExpr(Type sig_type) {
+  u32 exn_stack_height;
+  CHECK_RESULT(
+      validator_.GetCatchCount(label_stack_.size() - 1, &exn_stack_height));
+  u32 value_stack_height = validator_.type_stack_size();
+  CHECK_RESULT(validator_.OnTry(loc, sig_type));
+  // Push a label that tracks mapping of exn -> catch
+  PushLabel(LabelKind::Try, Istream::kInvalidOffset, Istream::kInvalidOffset,
+            func_->handlers.size());
+  func_->handlers.push_back(HandlerDesc{HandlerKind::Catch,
+                                        istream_.end(),
+                                        Istream::kInvalidOffset,
+                                        {},
+                                        {Istream::kInvalidOffset},
+                                        value_stack_height,
+                                        exn_stack_height});
+  return Result::Ok;
+}
+
+Result BinaryReaderInterp::OnCatchExpr(Index tag_index) {
+  CHECK_RESULT(validator_.OnCatch(loc, Var(tag_index), false));
+  Label* label = TopLabel();
+  HandlerDesc& desc = func_->handlers[label->handler_desc_index];
+  desc.kind = HandlerKind::Catch;
+  // Jump to the end of the block at the end of the previous try or catch.
+  if (label->kind == LabelKind::Block) {
+    istream_.EmitCatchDrop(1);
+  }
```

Comment:
Shoudn't we do this even when there's only a single `catch` or `catch_all`?

Reply:
If there's only a single `catch`, it should get handled by the code path in `OnEnd` here: https://github.com/WebAssembly/wabt/blob/d9cf828520103f916083af2e96aff914800c81f5/src/interp/binary-reader-interp.cc#L1114

(because there's no next catch block that would run this cleanup)

Reply:
```wasm
try
  ...
catch $e0  ;; This gets dropped by this fix
  ...
catch $e1  ;; Where does this get dropped?
  ...
catch $e2  ;; This gets dropped by `OnEnd`
  ...
end
```

Reply:
The second one should get dropped by this fix too, because each `catch` opcode triggers an `OnCatchExpr`call . On the first `OnCatchExpr` it doesn't drop anything as the previous code is for the try body (`label->kind == LabelKind::Try`). On the second, it drops `$e0`. On the third, it drops `$e1`. Finally, `OnEndExpr` drops `$e2`.

Reply:
Sorry again, how about adding one-line comment saying this is dropping the previous catch's exception then?

Reply:
Sure no problem, added.